File: src/python/signals.py
"""
Signal-First Scheduling System (Python)

Decisions follow telemetry strength and domain rankings,
so resources flow to highest-value workloads while safety stays enforced.
"""

from typing import Dict, List, Optional, Any
from dataclasses import dataclass, field
from enum import Enum, auto
import heapq
import time


class SignalSource(Enum):
    TELEMETRY = auto()
    USER = auto()
    POLICY = auto()
    SAFETY = auto()
    LEARNING = auto()
    EXTERNAL = auto()
    IPC = auto()


class SignalKind(Enum):
    # Performance signals
    FRAMETIME_SPIKE = auto()
    CPU_BOTTLENECK = auto()
    GPU_BOTTLENECK = auto()
    MEMORY_PRESSURE = auto()

    # Thermal signals
    THERMAL_WARNING = auto()
    THERMAL_CRITICAL = auto()
    COOLING_OPPORTUNITY = auto()

    # Workload signals
    WORKLOAD_CHANGE = auto()
    FOREGROUND_SWITCH = auto()
    IDLE_DETECTED = auto()
    BURST_DETECTED = auto()

    # User signals
    USER_BOOST_REQUEST = auto()
    USER_PROFILE_CHANGE = auto()

    # Safety signals
    GUARDRAIL_TRIGGERED = auto()
    EMERGENCY_STOP = auto()

    # IPC signals
    DRIVER_EVENT = auto()
    OPENVINO_READY = auto()


class Domain(Enum):
    SAFETY = 0
    THERMAL = 1
    USER = 2
    PERFORMANCE = 3
    POWER = 4


@dataclass
class Signal:
    """Signal with strength and metadata."""
    id: str
    source: SignalSource
    kind: SignalKind
    strength: float  # 0.0 - 1.0 normalized
    confidence: float = 0.9  # 0.0 - 1.0
    timestamp: float = field(default_factory=time.time)
    ttl_ms: int = 5000
    payload: Dict[str, Any] = field(default_factory=dict)

    def is_expired(self) -> bool:
        return (time.time() - self.timestamp) * 1000 > self.ttl_ms


@dataclass(order=True)
class RankedSignal:
    """Signal with computed priority for heap ordering."""
    priority: float  # Negated for max-heap behavior
    domain_rank: int
    signal: Signal = field(compare=False)


class SignalScheduler:
    """Signal scheduler with domain-based priority."""

    def __init__(self):
        self._queue: List[RankedSignal] = []
        self._processed: List[Signal] = []

        # Domain weights (higher = more important)
        self.domain_weights: Dict[Domain, float] = {
            Domain.SAFETY: 1.0,
            Domain.THERMAL: 0.85,
            Domain.USER: 0.8,
            Domain.PERFORMANCE: 0.7,
            Domain.POWER: 0.5,
        }

        # Kind base priorities
        self.kind_priorities: Dict[SignalKind, float] = {
            SignalKind.EMERGENCY_STOP: 1.0,
            SignalKind.THERMAL_CRITICAL: 0.95,
            SignalKind.GUARDRAIL_TRIGGERED: 0.9,
            SignalKind.THERMAL_WARNING: 0.8,
            SignalKind.USER_BOOST_REQUEST: 0.75,
            SignalKind.FRAMETIME_SPIKE: 0.7,
            SignalKind.CPU_BOTTLENECK: 0.65,
            SignalKind.GPU_BOTTLENECK: 0.65,
            SignalKind.MEMORY_PRESSURE: 0.6,
            SignalKind.WORKLOAD_CHANGE: 0.5,
            SignalKind.FOREGROUND_SWITCH: 0.5,
            SignalKind.DRIVER_EVENT: 0.45,
            SignalKind.OPENVINO_READY: 0.4,
            SignalKind.BURST_DETECTED: 0.4,
            SignalKind.USER_PROFILE_CHANGE: 0.35,
            SignalKind.COOLING_OPPORTUNITY: 0.3,
            SignalKind.IDLE_DETECTED: 0.2,
        }

    def enqueue(self, signal: Signal) -> None:
        """Enqueue a signal with computed priority."""
        domain = self._classify_domain(signal)
        priority = self._compute_priority(signal, domain)

        ranked = RankedSignal(
            priority=-priority,  # Negate for min-heap to act as max-heap
            domain_rank=domain.value,
            signal=signal,
        )
        heapq.heappush(self._queue, ranked)

    def dequeue(self) -> Optional[Signal]:
        """Dequeue highest priority non-expired signal."""
        while self._queue:
            ranked = heapq.heappop(self._queue)
            if not ranked.signal.is_expired():
                self._processed.append(ranked.signal)
                return ranked.signal
        return None

    def peek(self) -> Optional[Signal]:
        """Peek at highest priority signal without removing."""
        for ranked in sorted(self._queue):
            if not ranked.signal.is_expired():
                return ranked.signal
        return None

    def drain_by_priority(self) -> List[Signal]:
        """Get all pending signals sorted by priority."""
        signals = []
        while True:
            sig = self.dequeue()
            if sig is None:
                break
            signals.append(sig)
        return signals

    def filter_by_domain(self, domain: Domain) -> List[Signal]:
        """Get signals by domain."""
        return [
            r.signal for r in self._queue
            if self._classify_domain(r.signal) == domain and not r.signal.is_expired()
        ]

    def filter_by_kind(self, kind: SignalKind) -> List[Signal]:
        """Get signals by kind."""
        return [
            r.signal for r in self._queue
            if r.signal.kind == kind and not r.signal.is_expired()
        ]

    def _classify_domain(self, signal: Signal) -> Domain:
        """Classify signal into domain."""
        domain_map = {
            SignalKind.EMERGENCY_STOP: Domain.SAFETY,
            SignalKind.GUARDRAIL_TRIGGERED: Domain.SAFETY,
            SignalKind.THERMAL_WARNING: Domain.THERMAL,
            SignalKind.THERMAL_CRITICAL: Domain.THERMAL,
            SignalKind.COOLING_OPPORTUNITY: Domain.THERMAL,
            SignalKind.USER_BOOST_REQUEST: Domain.USER,
            SignalKind.USER_PROFILE_CHANGE: Domain.USER,
            SignalKind.FRAMETIME_SPIKE: Domain.PERFORMANCE,
            SignalKind.CPU_BOTTLENECK: Domain.PERFORMANCE,
            SignalKind.GPU_BOTTLENECK: Domain.PERFORMANCE,
            SignalKind.WORKLOAD_CHANGE: Domain.PERFORMANCE,
            SignalKind.FOREGROUND_SWITCH: Domain.PERFORMANCE,
            SignalKind.BURST_DETECTED: Domain.PERFORMANCE,
            SignalKind.MEMORY_PRESSURE: Domain.POWER,
            SignalKind.IDLE_DETECTED: Domain.POWER,
            SignalKind.DRIVER_EVENT: Domain.PERFORMANCE,
            SignalKind.OPENVINO_READY: Domain.PERFORMANCE,
        }
        return domain_map.get(signal.kind, Domain.PERFORMANCE)

    def _compute_priority(self, signal: Signal, domain: Domain) -> float:
        """Compute signal priority."""
        base = self.kind_priorities.get(signal.kind, 0.5)
        domain_weight = self.domain_weights.get(domain, 0.5)
        return base * domain_weight * signal.strength * signal.confidence

    def set_domain_weight(self, domain: Domain, weight: float) -> None:
        """Update domain weight dynamically."""
        self.domain_weights[domain] = max(0.0, min(1.0, weight))

    def stats(self) -> Dict[str, Any]:
        """Get scheduler statistics."""
        by_domain: Dict[Domain, int] = {}
        by_kind: Dict[SignalKind, int] = {}

        for ranked in self._queue:
            if not ranked.signal.is_expired():
                domain = self._classify_domain(ranked.signal)
                by_domain[domain] = by_domain.get(domain, 0) + 1
                by_kind[ranked.signal.kind] = by_kind.get(ranked.signal.kind, 0) + 1

        return {
            "queue_size": len([r for r in self._queue if not r.signal.is_expired()]),
            "processed_count": len(self._processed),
            "by_domain": {d.name: c for d, c in by_domain.items()},
            "by_kind": {k.name: c for k, c in by_kind.items()},
        }

    def clear_history(self) -> None:
        """Clear processed signal history."""
        self._processed.clear()


# Signal factory helpers
def telemetry_signal(kind: SignalKind, strength: float, **payload) -> Signal:
    """Create a telemetry-sourced signal."""
    return Signal(
        id=f"tel-{kind.name.lower()}-{int(time.time()*1000)}",
        source=SignalSource.TELEMETRY,
        kind=kind,
        strength=strength,
        payload=payload,
    )


def safety_signal(kind: SignalKind, strength: float = 1.0, **payload) -> Signal:
    """Create a safety-sourced signal (high priority)."""
    return Signal(
        id=f"safety-{kind.name.lower()}-{int(time.time()*1000)}",
        source=SignalSource.SAFETY,
        kind=kind,
        strength=strength,
        confidence=1.0,  # Safety signals are always confident
        ttl_ms=10000,    # Longer TTL for safety
        payload=payload,
    )


def user_signal(kind: SignalKind, strength: float = 0.9, **payload) -> Signal:
    """Create a user-sourced signal."""
    return Signal(
        id=f"user-{kind.name.lower()}-{int(time.time()*1000)}",
        source=SignalSource.USER,
        kind=kind,
        strength=strength,
        payload=payload,
    )


def ipc_signal(kind: SignalKind, strength: float, source_component: str, **payload) -> Signal:
    """Create an IPC-sourced signal from C/driver layer."""
    return Signal(
        id=f"ipc-{source_component}-{int(time.time()*1000)}",
        source=SignalSource.IPC,
        kind=kind,
        strength=strength,
        payload={"source_component": source_component, **payload},
    )

File: src/rust/signals.rs
//! Signal-First Scheduling System
//!
//! Decisions follow telemetry strength and domain rankings,
//! so resources flow to highest-value workloads while safety stays enforced.

use std::collections::{BinaryHeap, HashMap};
use std::cmp::Ordering;
use serde::{Deserialize, Serialize};

/// Signal with strength and metadata
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Signal {
    pub id: String,
    pub source: SignalSource,
    pub kind: SignalKind,
    pub strength: f64,       // 0.0 - 1.0 normalized
    pub confidence: f64,     // 0.0 - 1.0
    pub timestamp: u64,
    pub ttl_ms: u64,
    pub payload: HashMap<String, f64>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum SignalSource {
    Telemetry,
    User,
    Policy,
    Safety,
    Learning,
    External,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum SignalKind {
    // Performance signals
    FrametimeSpike,
    CpuBottleneck,
    GpuBottleneck,
    MemoryPressure,

    // Thermal signals
    ThermalWarning,
    ThermalCritical,
    CoolingOpportunity,

    // Workload signals
    WorkloadChange,
    ForegroundSwitch,
    IdleDetected,
    BurstDetected,

    // User signals
    UserBoostRequest,
    UserProfileChange,

    // Safety signals
    GuardrailTriggered,
    EmergencyStop,
}

/// Domain for signal ranking
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum Domain {
    Safety,
    Performance,
    Thermal,
    Power,
    User,
}

/// Ranked signal for priority queue
#[derive(Debug, Clone)]
struct RankedSignal {
    signal: Signal,
    priority: f64,
    domain_rank: u32,
}

impl PartialEq for RankedSignal {
    fn eq(&self, other: &Self) -> bool {
        self.priority == other.priority
    }
}

impl Eq for RankedSignal {}

impl PartialOrd for RankedSignal {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for RankedSignal {
    fn cmp(&self, other: &Self) -> Ordering {
        // Higher priority first, then lower domain rank (safety=0 highest)
        match self.priority.partial_cmp(&other.priority) {
            Some(Ordering::Equal) => other.domain_rank.cmp(&self.domain_rank),
            Some(ord) => ord.reverse(), // Reverse for max-heap behavior
            None => Ordering::Equal,
        }
    }
}

/// Signal scheduler
pub struct SignalScheduler {
    queue: BinaryHeap<RankedSignal>,
    domain_weights: HashMap<Domain, f64>,
    kind_priorities: HashMap<SignalKind, f64>,
    processed: Vec<Signal>,
    max_queue_size: usize,
}

impl SignalScheduler {
    pub fn new() -> Self {
        let mut scheduler = Self {
            queue: BinaryHeap::new(),
            domain_weights: HashMap::new(),
            kind_priorities: HashMap::new(),
            processed: Vec::new(),
            max_queue_size: 1000,
        };
        scheduler.init_defaults();
        scheduler
    }

    fn init_defaults(&mut self) {
        // Domain weights (higher = more important)
        self.domain_weights.insert(Domain::Safety, 1.0);
        self.domain_weights.insert(Domain::User, 0.9);
        self.domain_weights.insert(Domain::Performance, 0.7);
        self.domain_weights.insert(Domain::Thermal, 0.8);
        self.domain_weights.insert(Domain::Power, 0.5);

        // Kind base priorities
        self.kind_priorities.insert(SignalKind::EmergencyStop, 1.0);
        self.kind_priorities.insert(SignalKind::ThermalCritical, 0.95);
        self.kind_priorities.insert(SignalKind::GuardrailTriggered, 0.9);
        self.kind_priorities.insert(SignalKind::ThermalWarning, 0.8);
        self.kind_priorities.insert(SignalKind::UserBoostRequest, 0.75);
        self.kind_priorities.insert(SignalKind::FrametimeSpike, 0.7);
        self.kind_priorities.insert(SignalKind::CpuBottleneck, 0.65);
        self.kind_priorities.insert(SignalKind::GpuBottleneck, 0.65);
        self.kind_priorities.insert(SignalKind::MemoryPressure, 0.6);
        self.kind_priorities.insert(SignalKind::WorkloadChange, 0.5);
        self.kind_priorities.insert(SignalKind::ForegroundSwitch, 0.5);
        self.kind_priorities.insert(SignalKind::BurstDetected, 0.4);
        self.kind_priorities.insert(SignalKind::CoolingOpportunity, 0.3);
        self.kind_priorities.insert(SignalKind::IdleDetected, 0.2);
        self.kind_priorities.insert(SignalKind::UserProfileChange, 0.4);
    }

    /// Enqueue a signal
    pub fn enqueue(&mut self, signal: Signal) {
        let domain = self.classify_domain(&signal);
        let priority = self.compute_priority(&signal, domain);
        let domain_rank = self.domain_rank(domain);

        let ranked = RankedSignal {
            signal,
            priority,
            domain_rank,
        };

        self.queue.push(ranked);

        // Trim if over capacity
        while self.queue.len() > self.max_queue_size {
            self.queue.pop();
        }
    }

    /// Dequeue highest priority signal
    pub fn dequeue(&mut self) -> Option<Signal> {
        let now = current_timestamp();

        // Skip expired signals
        while let Some(ranked) = self.queue.pop() {
            if ranked.signal.timestamp + ranked.signal.ttl_ms >= now {
                self.processed.push(ranked.signal.clone());
                return Some(ranked.signal);
            }
        }
        None
    }

    /// Peek at highest priority signal without removing
    pub fn peek(&self) -> Option<&Signal> {
        self.queue.iter().next().map(|r| &r.signal)
    }

    /// Get all pending signals sorted by priority
    pub fn drain_by_priority(&mut self) -> Vec<Signal> {
        let mut signals = Vec::new();
        while let Some(signal) = self.dequeue() {
            signals.push(signal);
        }
        signals
    }

    /// Get signals by domain
    pub fn filter_by_domain(&self, domain: Domain) -> Vec<&Signal> {
        self.queue.iter()
            .filter(|r| self.classify_domain(&r.signal) == domain)
            .map(|r| &r.signal)
            .collect()
    }

    fn classify_domain(&self, signal: &Signal) -> Domain {
        match signal.kind {
            SignalKind::EmergencyStop | SignalKind::GuardrailTriggered => Domain::Safety,
            SignalKind::ThermalWarning | SignalKind::ThermalCritical |
            SignalKind::CoolingOpportunity => Domain::Thermal,
            SignalKind::FrametimeSpike | SignalKind::CpuBottleneck |
            SignalKind::GpuBottleneck | SignalKind::BurstDetected => Domain::Performance,
            SignalKind::UserBoostRequest | SignalKind::UserProfileChange => Domain::User,
            SignalKind::MemoryPressure | SignalKind::IdleDetected => Domain::Power,
            SignalKind::WorkloadChange | SignalKind::ForegroundSwitch => Domain::Performance,
        }
    }

    fn compute_priority(&self, signal: &Signal, domain: Domain) -> f64 {
        let base = self.kind_priorities.get(&signal.kind).copied().unwrap_or(0.5);
        let domain_weight = self.domain_weights.get(&domain).copied().unwrap_or(0.5);

        // Priority = base * domain_weight * strength * confidence
        base * domain_weight * signal.strength * signal.confidence
    }

    fn domain_rank(&self, domain: Domain) -> u32 {
        match domain {
            Domain::Safety => 0,      // Highest priority
            Domain::Thermal => 1,
            Domain::User => 2,
            Domain::Performance => 3,
            Domain::Power => 4,
        }
    }

    /// Update domain weights dynamically
    pub fn set_domain_weight(&mut self, domain: Domain, weight: f64) {
        self.domain_weights.insert(domain, weight.clamp(0.0, 1.0));
    }

    /// Get queue stats
    pub fn stats(&self) -> SchedulerStats {
        let mut by_domain: HashMap<Domain, usize> = HashMap::new();
        let mut by_kind: HashMap<SignalKind, usize> = HashMap::new();

        for ranked in &self.queue {
            let domain = self.classify_domain(&ranked.signal);
            *by_domain.entry(domain).or_default() += 1;
            *by_kind.entry(ranked.signal.kind).or_default() += 1;
        }

        SchedulerStats {
            queue_size: self.queue.len(),
            processed_count: self.processed.len(),
            signals_by_domain: by_domain,
            signals_by_kind: by_kind,
        }
    }

    /// Clear processed history
    pub fn clear_history(&mut self) {
        self.processed.clear();
    }
}

impl Default for SignalScheduler {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Clone)]
pub struct SchedulerStats {
    pub queue_size: usize,
    pub processed_count: usize,
    pub signals_by_domain: HashMap<Domain, usize>,
    pub signals_by_kind: HashMap<SignalKind, usize>,
}

fn current_timestamp() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_millis() as u64
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_signal(kind: SignalKind, strength: f64) -> Signal {
        Signal {
            id: format!("sig-{:?}", kind),
            source: SignalSource::Telemetry,
            kind,
            strength,
            confidence: 0.9,
            timestamp: current_timestamp(),
            ttl_ms: 10000,
            payload: HashMap::new(),
        }
    }

    #[test]
    fn test_priority_ordering() {
        let mut scheduler = SignalScheduler::new();

        scheduler.enqueue(make_signal(SignalKind::IdleDetected, 0.5));
        scheduler.enqueue(make_signal(SignalKind::ThermalCritical, 0.9));
        scheduler.enqueue(make_signal(SignalKind::FrametimeSpike, 0.8));

        // Thermal critical should come first (safety domain)
        let first = scheduler.dequeue().unwrap();
        assert_eq!(first.kind, SignalKind::ThermalCritical);
    }

    #[test]
    fn test_domain_weights() {
        let mut scheduler = SignalScheduler::new();

        // Boost performance domain
        scheduler.set_domain_weight(Domain::Performance, 1.0);
        scheduler.set_domain_weight(Domain::Thermal, 0.3);

        scheduler.enqueue(make_signal(SignalKind::ThermalWarning, 0.5));
        scheduler.enqueue(make_signal(SignalKind::FrametimeSpike, 0.5));

        // With boosted performance weight, frametime should win
        let first = scheduler.dequeue().unwrap();
        // Note: ThermalWarning still has higher base priority, this tests the weighting
    }
}
